Html css popup

dmytro-botnarenko opened the pull request:
### html-css-popup

[Demo](https://dmytro-botnarenko.github.io/hooli-style-popup) |
[Code base](https://github.com/dmytro-botnarenko/hooli-style-popup)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

kasionio commented:
Nice job! But it does not satisfy the task description requirements. Please recheck it or check the `common mistakes` link

dmytro-botnarenko commented:
Working on it...

dmytro-botnarenko commented:
Hi! I've updated my project files.
Main changes in HTML (accordingly to common mistakes):
- `nav` tag now has `ul>li` nested structure
- alt text changed to more directive meaning
Main changes in CSS:
- changes are associated with a bit new HTML structure;
- changes at `:focus` states (now popup/more/less) 'buttons' zones has focus from keyboard;

Known issue:
Focus order on the "Less" button are ahead of focus'es on the icons in expanding Popup part. This is due to the fact that the same checkbox (`.toggle__more`) that is above them (icons) in the document row is actually focused. But in such a structure, I do not yet know how it can be changed.(

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+.toggle__more:checked ~ .label__toggle__more {
+    display: none;
+}
+.toggle__more:checked {
```
try to remove this property and look how it will work

dmytro-botnarenko replied:
I made this style because when the popup is fully expanded, when you switch from the keyboard, the focus falls to the old position of the `.toggle__more` checkbox. Therefore, if I remove this property, the `.toggle__more` checkbox will have it's initial position (not at the bottom of the popup window) and when focusing, it will look like this: [![https://imgur.com/v2eJCgG.png](https://imgur.com/v2eJCgG.png)](https://imgur.com/v2eJCgG.png)

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+    box-shadow: 0px 2px 2px 1px rgba(0, 0, 0, 0.2);
+    border-radius: 3px;
+    cursor: pointer;
+    z-index: 1;
```
z-indexes looks redundant here

dmytro-botnarenko replied:
I added this `z-index` so that the label is always highlighted when hovering over it. If you remove it, the label will light up only on its edges, but when you get to the input checkbox zone (which is higher in HTML structure), the label hovering effect will disappear.(

kasionio reviewed (changes requested):
Good job! You can take a look at the video on how to use correctly visually hide checbox [video](https://www.youtube.com/watch?v=E6kLaaQFctU&t=413s&ab_channel=%D0%92%D0%B0%D0%B4%D0%B8%D0%BC%D0%9C%D0%B0%D0%BA%D0%B5%D0%B5%D0%B2)

Also some unexpected background appears while using keyboard

Please also take a look at the comments below

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/README.md:
```diff
@@ -0,0 +1,10 @@
+# kottans-html-css-popup
```
Changed the project name to "hooli-style-popup" and added a demo link on github pages.

kasionio replied:
You should delete Readme file, it is not needed here

dmytro-botnarenko replied:
My bad. Didn't understand from the very beginning. Deleted.

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/index.html:
```diff
@@ -0,0 +1,139 @@
+    <meta http-equiv="X-UA-Compatible" content="IE=edge">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <link rel="stylesheet" type="text/css" href="style.css">
+    <title>Document</title>
```
Changed `title` to "Hooli-style Popup"

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/index.html:
```diff
@@ -0,0 +1,139 @@
+                    </li>
+                    <li class="navigation__list">                
+                        <input id="toggle" class="toggle" type="checkbox">
+                        <label for="toggle" class="navigation__link"><img class="navigation__image" src="img/menu.svg" alt="Go to popup menu"></label>
```
Thank you for this useful link. This is not the first article in which I read about the `alt` text.)

In the case of a specific line (23), I can only assume that it is maybe better to write *"Open menu with other links"* or *"Open menu with additional links"* instead of a bit abstract *"Go to popup menu"*. Does this wording fits better in this situation? For now, I'll make a correction and replace with *"Open menu with additional links"*. All other `alt` text in main navigation strip now has same pattern *"Open ..."*

As for the `alt` texts for the icons in the popup menu itself, this is not the first time I have come across that you **may** not to write `alt` text to them if there is a text signature in the link. I've added `alt` text to the icons in Popup menu, but maybe they are redundant. But [this article](https://axesslab.com/alt-texts/) convinced me, and I decided to remove the `alt` texts, since there is the necessary similar text in the `a` links (containers) of the images.

I hope I have interpreted these standards correctly now.

kasionio replied:
Yeah, right decision 👍

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+    outline: none;
+    border-radius: 3px;
+    background-color: rgba(0, 0, 0, 0.1);
+    transition: background-color 1s;
```
Removed `transition:`

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+    width: 35%;
+    height: 35%;
+    border-radius: 15%;
+    align-self: center;
```
Yep. Removed `align-self:`

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+    position: absolute;
+    top: 450px;    
+}
+::-webkit-scrollbar {
```
As far as I understand, if I want to visually remove the scrollbar, leaving its functionality, I have next [way](https://www.w3schools.com/howto/howto_css_hide_scrollbars.asp#:~:text=Hide%20Scrollbars%20But%20Keep%20Functionality).

`::-webkit-scrollbar` works in Chrome (our main browser by this task, and in other browsers, except FF & IE). 

`scrollbar-width` - a "standardized method of styling scrollbars" from [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-width) works only in FF. Thus, I have two options: either abandon the idea of ​​visually hiding the scrollbar or make something like w3schools proposed.

BTW. Modern Firefox has great styled standard scrollbar with hiding effect and my popup window looks looks just as good, but when hovering, there is a non-distracting neat scrollbar. If Сhrome had the same, and not the its ugly standard one, then I probably wouldn't do anything. 

So, I've made like this:
``` css
.popup__container::-webkit-scrollbar {width: 0px;} /*Chrome, Safari, Opera, Edge*/
.popup__container {
    -ms-overflow-style: none;  /*IE*/
    scrollbar-width: auto;  /*Firefox*/}
```

kasionio replied:
Nice!

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+.toggle__more:checked ~ .label__toggle__more ~ .popup__more {
+    display:grid;
+    grid-template-columns: repeat(3,1fr);
+    grid-auto-flow: row;
```
Removed `grid-auto-flow: row;`

dmytro-botnarenko commented:
> Good job! You can take a look at the video on how to use correctly visually hide checbox [video](https://www.youtube.com/watch?v=E6kLaaQFctU&t=413s&ab_channel=%D0%92%D0%B0%D0%B4%D0%B8%D0%BC%D0%9C%D0%B0%D0%BA%D0%B5%D0%B5%D0%B2)

Here I did not understand a little, because in the code I used the `appearance: none` (lines 73-76) property that was noted in this video as the optimal choice.
> 
> Also some unexpected background appears while using keyboard ![image](https://user-images.githubusercontent.com/21179601/185766000-db3ba52c-7e75-4a0d-bec3-2e6b3857dafb.png)

It was a focus area and I've changed it now to outline style.

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,164 @@
+    letter-spacing: 0.05em;
+}
+a {
+	color: #794f45;
```
Done with Prettier

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,153 @@
+  background-color: white;
+  font-family: Tahoma, Helvetica, Arial, sans-serif;
+  color: #fff;
+  font-size: calc(12px + 0.390625vw);
```
why such a strange number?

dmytro-botnarenko replied:
I used [this article](https://matthewjamestaylor.com/responsive-font-size#:~:text=Fluid%20Responsive%20Font%20Size) to implement responsive `font-size`. It is possible that this is a very subjective opinion. Now I've changed it to `font-size: calc(15px + 0.39vw)`. Is this ok?

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,153 @@
+  box-shadow: 0px 1px 2px 1px rgba(0, 0, 0, 0.2);
+}
+.wrapper {
+  min-height: 100%;
```
why this property needed?

dmytro-botnarenko replied:
At the beginning of the project I needed this block, but now I realized that it seems that this `div .wrapper` is not needed at all. Removed in HTML and CSS.

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/index.html:
```diff
@@ -0,0 +1,215 @@
+      <header class="header">
+        <nav>
+          <ul class="navigation__menu">
+            <li class="navigation__list">
```
li - list item. Class `navigation__list` seems not correct naming

dmytro-botnarenko replied:
Made `navigation__item`. Now it's `navigation__menu` -> `navigation__item` -> `navigation__link` chain. Is it ok?

kasionio replied:
Nice, I like it:) If you are using BEM it should have nesting structure like nav-menu nav-menu_item nav-menu_item_link

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,153 @@
+.toggle:focus ~ .label__toggle,
+.toggle__more:focus ~ .label__toggle__more,
+.toggle__more:focus ~ .label__toggle__less {
+  outline: 2px solid rgba(0, 0, 0, 0.1);
```
why do you need rgba here? is there any reason to use alpha?

dmytro-botnarenko replied:
Simplified the color to `rgb (200, 200, 200)` here and in some other classes.

kasionio reviewed (changes requested):
Good job! But one of the main task requirements is not implemented yet - I can't reach `Less` button from the keyboard. And one suggestion - on my screen top border looks weird, it is almost not visible, also no gap between the element container and button. Please take a look at the screenshot below.
![image](https://user-images.githubusercontent.com/21179601/186003492-ee5fdd57-3430-4df0-928a-6de914f9ee78.png)

dmytro-botnarenko commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,153 @@
+}
+.popup__image {
+  height: 35px;
+  width: 48px;
```
Yes, `object-fit: contain` and different width/height ). Removed it.

dmytro-botnarenko commented:
> Good job! But one of the main task requirements is not implemented yet - I can't reach Less button from the keyboard. And one suggestion - on my screen top border looks weird, it is almost not visible, also no gap between the element container and button. Please take a look at the screenshot below.

This issue is known to me, but i cannot solve it, unfortunately. The Less button is reachable from the keyboard, but the order of the focusing is a bit wrong. It comes after "main" popup icons (after "Documents" icon) and before expanding zone. This happens because the focus is activated by the `input` that goes in the HTML earlier than the lower icons and it is impossible to move it without losing control of the elements.
The "less" button was not in the task and it can be removed, just the expanded window then cannot be collapsed. I tried to do this option, but I can't solve this focusing order.

> And one suggestion - on my screen top border looks weird, it is almost not visible, also no gap between the element container and button. Please take a look at the screenshot below.

I've changed a bit `box-shadow` in popup window and in some other elements for similarity. Plus I've added `margin` to the top of labels and now it seems that gap between container and button is sufficient.

dmytro-botnarenko commented:
> I believe you can! Struggling is also a part of a developer's life ;) You can find a decision in the works of other students or just googling if you are stuck. If you won't succeed in a couple of days let me know and I'll help, but try to solve it on your own.

HelIo! I went through quite a few examples from other students merged PRs. Most of their works do not have a "less" button and are made either with a button-link to Kottans site, or without it at all. I did find a few pages that have a "less" button, here they are: [one](https://vovan-zt.github.io/popup2022/) and [two](https://olegpopovych.github.io/html_css_popup/). But they all have the same focus queue problem. Unfortunately, the googling gives a lot of options for something similar, but only similar.

I tried to make these two switches on two radio buttons (with the same `name` attribute, and "Less" radio on the bottom of the container with `label` Less), but I did not take into account that the focus through the tab works only with the entire radio cluster, appears only on the first input, and inside it can be switched only with arrows, not with tab.

Then I made something that could be called a 'crutch'. I put a queue through `tabindex`. Focusing now works as it should. I did not prescribe `tabindexes` to any elements that are not acceptable for this, everything is according to the rules, the focus goes only to the necessary active/link/input elements. 

Yes, i understand this is not the best solution but it works in my case. Unfortunately I can't find a better one on pure css. Another option is to either transform the "less" button and make it a simple link to the kottans site, as most do, or make something like [this](https://yasya23.github.io/Hooli-style-Popup/), where "more/less" button is the same button in one position.

kasionio commented:
I'll take a look later how I can help you to handle it. In a couple of days

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,154 @@
+  box-shadow: 0 0 2px 2px rgba(0, 0, 0, 0.2);
+}
+.navigation__image {
+  width: 27%;
```
this approach is not working well for small screens. icons are too small
![image](https://user-images.githubusercontent.com/21179601/187511978-887ba5b0-3644-4230-a51b-99e021fc8391.png)

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,154 @@
+  display: flex;
+  justify-content: center;
+  padding: 2px 5px;
+  min-height: 30px;
```
seems `min-height` is not needed

kasionio commented on submissions/dmytro-botnarenko/html-css-popup/style.css:
```diff
@@ -0,0 +1,154 @@
+}
+.popup__image {
+  height: 35px;
+  object-fit: contain;
```
seems it is not have effect here

kasionio reviewed (approved):
It is almost working well, nice job! This task is more about [graceful degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation) and some tricks which we can use without JS. It is fair enough, let's move forward.

dmytro-botnarenko commented:
> It is almost working well, nice job! This task is more about [graceful degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation) and some tricks which we can use without JS. It is fair enough, let's move forward.

Thank you for your attention and advices on my project! :wink:
